### PR TITLE
Validate IP address and port of received packets (take 2)

### PR DESCRIPTION
**Pull request by acd**
Verify peer IP and port prior to handling received packets.  I didn't rework the logic as proposed in #22 as it would assign the tid value prior to parsing the received packet (which could be invalid).

### CODE REVIEW

**Comment by pin** [comment]
Regarding, logic change - that's right, makes sense. Sorry for misleading you!

Was digging into test failure, that's what I've found:

Client fails on IP address comparison on receive on linux. The reason it fails is that on linux server start listening to [::]:NNNN (all intefaces) _and_ uses that address as a server address when initializing a client. That happen in `makeTestServer` function in `tftp_test.go`.

But then client receives a packet from [::1]:NNNN (lo interface) and IP comparison (`[::]` vs `[::1]`) fails.

On OS X that works because `localhost` is explicitly used (which resolves to `127.0.0.1` or something like that).

I've fixed the tests by explicitly choosing loopback interface to run tests. It is a good catch actually, I think it will be possible to remove hack for tests on OS X (`conn_darwin.go`). Yay!

I believe if you rebase on top of that tests will pass. They pass locally on my linux box.

But _should_ IP address be checked in the first place? So far the only data I have is the reference to [TFTP Multicast Option RFC](https://tools.ietf.org/html/rfc2090) in [atftp source](https://github.com/seveas/atftp/blob/e60ad90917c727b607975c4077b5b4e1bc85b316/tftp_file.c#L259). Note though, in that code IP check is only for multicast transfers!

Can't find any reference in other docs regarding IP address check. It seems logical: we have to receive the packet from specific address. In the other hand I'm worrying that I do not know some edge-cases where it could brake things. What do you think?

Not trying to dismiss the fact that unit-tests are little bit broken now :)

**Comment by pin** [comment]
It works!

Suggest the following way forward with this change:

Main question: should IP address be validated? No question TID should be of course. Let me know your thoughts on that, I will review some other TFTP code meanwhile (will take some time).

Unit-test fix: I see an advantage in that approach https://github.com/pin/tftp/pull/26 that it potentially fixes OS X flaps with IPv6 addresses and does not assume that `lo` have either `127.0.0.1` or `[::1]` IP address (which is somewhat strong assumption from my point of view, but not a big of a deal because it just tests in the other hand). Have to test if https://github.com/pin/tftp/pull/26 really fixes OS X flap: will take some time. If not, https://github.com/pin/tftp/pull/23/commits/29d2b723487b09f8042a955eb8afedb2f7f78a11 is way to go probably.

**Comment by acd** [comment]
Re: IP address being validated, I can't think of a reason not to validate the IP.  I believe most TFTP server implementations use connected sockets to respond to the client which hoists the IP and port validation to the kernel.  I haven't looked at any TFTP client code, but it wouldn't surprise me if they use an unconnected UDP socket to send the initial request packet and then call connect upon receipt of a response from the server.

Re: Unit tests, 29d2b72 works on both OS X and Linux; I've tested it.  I can make another commit to remove conn_darwin.go if you'd like.  I think it's cleaner to establish a TCP connection on localhost and use the address selected by the kernel vs. enumerating and looping through interfaces, but I'm fine with whatever you'd like to do.  I can test #26 on OS X if you'd like.

**Comment by pin** [comment]
IP validation: let me dig into that a little bit more over the weekend, connected socket argument sounds reasonable.

`conn_darwin.go` removal: do not  do it now because I'm not sure OS X tests are fixed. It was strange flap that manifested like this: you run a test and sometimes (very rarely) it times-out. After banging my head for a while I've figured it happen when for some reason OS X allows TFTP start listening on the same port `syslogd` already listening. Some wired clash of IPv6 and IPv4 and maybe that buggy test. Better to leave it like this and than try to remove OS X hack later in separate change.

**Comment by pin** [comment]
I've found that you are 100% right regarding IP check: for instance in `atftp` there is a `connect` call in `tftp_receive_request`: https://github.com/seveas/atftp/blob/master/tftpd.c#L710

Which makes sense to me now.

Also I've found that both changes that select IP to call in tests more carefully fix OS X flaps. There are two small nits about code change, I will add inline comments.

File: receiver.go [hunk 1]
```diff
@@ -154,11 +155,14 @@ func (r *receiver) receiveDatagram(l int) (int, *net.UDPAddr, error) {
 		if err != nil {
 			return 0, nil, err
 		}
-		// TODO: compare addr here?
+		if !addr.IP.Equal(r.addr.IP) || (r.tid != 0 && addr.Port != r.tid) {
+			return 0, nil, err
```
Comment (pin) [inline receiver.go]: Probably instead of `return 0, nil, err` should be `continue`?

Because `continue` just does what kernel otherwise would do on connected socket: just drop the frame.
In the other hand `return 0, nil, err` is in fact `return 0, nil, nil` here.

File: conn.go [hunk 2]
```diff
@@ -3,9 +3,54 @@
 package tftp
 
 import "net"
+import "fmt"
+import "strconv"
 
 const udp = "udp"
 
+var localhost string = determineLocalhost()
```
Comment (pin) [inline conn.go]: `determineLocalhost` called not only at tests start, also at the start of any process that uses the package. It is little bit of a hassle to fix: Probably function and variable should be moved to `_test` (e.g. `tftp_test.go`) file, but in order to do that it is necessary to move `localSystem` there too, etc...

It could be done similar to https://github.com/pin/tftp/pull/26 to completely get rid of `conn.go` and `conn_darwin.go`. It should be OK, with this change _darwin workaround is not required anymore.

**Comment by pin** [comment]
`determineLocalhost` is not critical for this PR, can be moved later (just probably should be moved before next release).

File: receiver.go (hunk 1 above)
Comment (acd) [inline receiver.go]: Yep agreed, `continue` is best.

**Comment by pin** [comment]
LGTM!

This is awesome set of changes! :+1

It seems that there is a great time to cut a new release after merging this PR.
